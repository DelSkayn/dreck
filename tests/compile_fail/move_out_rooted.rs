use dreck::*;

pub struct Container<'gc, 'own>(Option<Gc<'gc, 'own, Container<'gc, 'own>>>);

unsafe impl<'gc, 'own> Trace<'own> for Container<'gc, 'own> {
    fn needs_trace() -> bool
    where
        Self: Sized,
    {
        true
    }

    fn trace<'t>(&self, trace: Tracer<'t,'own>) {
        self.0.trace(trace)
    }
}

unsafe impl<'from,'to,'own> Bound<'to> for Container<'from, 'own> {
    type Rebound = Container<'to,'own>;
}

fn main() {
    new_root!(owner, root);

    let mut container = Container(None);
    let ptr = root.add(container);
    container = Container(Some(ptr));
    let ptr = root.add(container);

    root!(&root, ptr);

    root.collect(owner);

    let v = ptr.borrow_mut(owner, &root).0.take().unwrap();
    root.collect_full(owner);

    assert!(v.borrow(owner).0.is_none());
}
