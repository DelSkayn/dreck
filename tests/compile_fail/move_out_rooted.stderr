error[E0502]: cannot borrow `*owner` as immutable because it is also borrowed as mutable
  --> tests/compile_fail/move_out_rooted.rs:35:23
   |
34 |     let v = ptr.borrow_mut(owner, &root).0.take().unwrap();
   |                            ----- mutable borrow occurs here
35 |     root.collect_full(owner);
   |                       ^^^^^ immutable borrow occurs here
36 |
37 |     assert!(v.borrow(owner).0.is_none());
   |             - mutable borrow later used here

error[E0502]: cannot borrow `*owner` as immutable because it is also borrowed as mutable
  --> tests/compile_fail/move_out_rooted.rs:37:22
   |
34 |     let v = ptr.borrow_mut(owner, &root).0.take().unwrap();
   |                            ----- mutable borrow occurs here
...
37 |     assert!(v.borrow(owner).0.is_none());
   |               ------ ^^^^^ immutable borrow occurs here
   |               |
   |               mutable borrow later used by call
